### ISSUE TITLE
same network structure but very different performance

### ISSUE THREAD

**Issue by goodtogood** [AlexeyAB/darknet#5531]
I'v trained two models based on yolov3-tiny_3l.cfg.
two models have same input size (608) and anchor (default)
the only difference is the num of class.

model A
```
darknet.exe detector demo ./A.data ./A.cfg A.weights A.mp4 -benchmark   111fps
darknet.exe detector demo ./A.data ./A.cfg A.weights A.mp4 -dont_show -ext_output  105fps
```

model B
```
darknet.exe detector demo ./B.data ./B.cfg B.weights B.mp4 -benchmark   47fps
darknet.exe detector demo ./B.data ./B.cfg B.weights B.mp4 -dont_show -ext_output  38fps
```

how come? Is this normal?
thanks!

**Comment by cong235** [comment]
same question

**Comment by AlexeyAB** [comment]
* Your link is broken
* What GPU do you use?
* How many classes are in your cfg-file?
* Attach your cfg-files
* Show such screenshot:
![image](https://user-images.githubusercontent.com/4096485/81295433-140c1480-9079-11ea-969d-b47569375c33.png)

**Comment by goodtogood** [comment]
- the code compiled is this [commit](https://github.com/AlexeyAB/darknet/commit/1f60ca57ad90f34a0559783656fa1ba98c61bb6d)

- tests are running on rtx 2070s

- A has 2 classes, B has 138 classes,  does it matter ? 

[A.cfg.txt](https://github.com/AlexeyAB/darknet/files/4597780/A.cfg.txt)

[B.cfg.txt](https://github.com/AlexeyAB/darknet/files/4597781/B.cfg.txt)

![s](https://user-images.githubusercontent.com/10088733/81386047-f60ce580-9146-11ea-99e2-c1a033fc48ef.jpg)

**Comment by AlexeyAB** [comment]
Yes, there is still bottleneck on CPU. I will fix it.

I added temporal fix: https://github.com/AlexeyAB/darknet/commit/ccafff912bd42728aa873be7b4c7add2366858a8

----

**Before fix:**

CPU Intel Core i7 6500K - RTX2070 Desktop - yolov3-tiny_3l.cfg (yours)

* A (2 classes) - 166 FPS (-benchmark) - 151 FPS (-dont_show)

* B (138 classes) - 81 FPS (-benchmark) - 75 FPS (-dont_show)

-----

**After fix:**

* A (2 classes) - 197 FPS (-benchmark) - 162 FPS (-dont_show)

* B (138 classes) - 125 FPS (-benchmark) - 119 FPS (-dont_show)

**Comment by goodtogood** [comment]
CPU  xeon e3 1231 v3    GPU rtx2070s

Model B does has an improvment after this fix.

before fix:
- A (2 classes) - 111 FPS (-benchmark) - 105 FPS (-dont_show)

- B (138 classes) - 47 FPS (-benchmark) - 38 FPS (-dont_show)

after fix:
- A (2 classes) -  112 FPS (-benchmark) - 108 FPS (-dont_show)

- B (138 classes) - 66.5 FPS (-benchmark) - 61.9 FPS (-dont_show)

but I noticed  fps is still much lower than yours.
the cpu usage is just about 30% when running the test.
Is there something I missed ?

**Comment by AlexeyAB** [comment]
* CPU xeon e3 1231 v3 - **3.40 GHz** https://ark.intel.com/content/www/us/en/ark/products/80910/intel-xeon-processor-e3-1231-v3-8m-cache-3-40-ghz.html
* CPU Intel Core i7 6500K - **4.00 GHz (+~20%)** 
 https://ark.intel.com/content/www/us/en/ark/products/88195/intel-core-i7-6700k-processor-8m-cache-up-to-4-20-ghz.html

----

https://en.wikipedia.org/wiki/List_of_Nvidia_graphics_processing_units#GeForce_20_series

* RTX 2070 - **7.4 TFLops**
* RTX 2060 super - **9 TFlops (+~20%)**

So my CPU is 20% faster, but you GPU is 20% faster. FPS should not differ much.

----

GPU-usage - 68%
CPU-usage - 33%

----

What CPU and GPU usage do you get for `B (138 classes) - 66.5 FPS (-benchmark) ` ?

----

Also do in MSVS:

1. open `\darknet.sln` -> (right click on project) -> Debugging -> Commands arguments: set there:
`detector demo ./B.data ./B.cfg B.weights B.mp4 -benchmark`
(without `darknet.exe`)

2.  In the top MSVS tab `Debug` ->Performance profiling -> Start

    * then wait for 30 sec and close console window.
    * unfold lines as in my screenshots
    * and do 2 such screenshots:


![image](https://user-images.githubusercontent.com/4096485/81402748-a6c0b800-913a-11ea-91b9-d1f458f3912c.png)

**detect_in_thread**
![image](https://user-images.githubusercontent.com/4096485/81402779-b4763d80-913a-11ea-9fdd-69440da7ddd0.png)

**Comment by AlexeyAB** [comment]
I added another fix: https://github.com/AlexeyAB/darknet/commit/da050367a8b174f751a2c9720195d51a369c4a4a

* A (2 classes) - 208 FPS (-benchmark) - 164 FPS (-dont_show)
* B (138 classes) - 139 FPS (-benchmark) - 122 FPS (-dont_show)

**Comment by goodtogood** [comment]
> What CPU and GPU usage do you get for B (138 classes) - 66.5 FPS (-benchmark) ?
![cpugpu](https://user-images.githubusercontent.com/10088733/81424715-f62dd500-9188-11ea-8927-c843034417b2.jpg)

> Performance profiling 

![p1](https://user-images.githubusercontent.com/10088733/81424774-0d6cc280-9189-11ea-8dee-2fed72dbf271.jpg)


![p2](https://user-images.githubusercontent.com/10088733/81424785-11004980-9189-11ea-94d7-1209804056a9.jpg)

**Comment by goodtogood** [comment]
after this fix  [da05036](https://github.com/AlexeyAB/darknet/commit/da050367a8b174f751a2c9720195d51a369c4a4a)

- A (2 classes) - 112 FPS (-benchmark) - 121 FPS (-dont_show)

- B (138 classes) - 66.5 FPS (-benchmark) - 63.1 FPS (-dont_show)

It seems the data is a little weird.

**Comment by AlexeyAB** [comment]
Did you set -banchmark or -dont_show there?

> 1. open `\darknet.sln` -> (right click on project) -> Debugging -> Commands arguments: set there:
> `detector demo ./B.data ./B.cfg B.weights B.mp4 -benchmark`

This is extremely strange if you get 12% CPU-time for `get_network_box()` when using the `-benchmark` flag:

![image](https://user-images.githubusercontent.com/4096485/81450962-2c208880-918c-11ea-83c8-f16ea2364a19.png)

**Comment by goodtogood** [comment]
Yes, I did set  `detector demo ./B.data ./B.cfg B.weights B.mp4 -benchmark`  as you said.

![p3](https://user-images.githubusercontent.com/10088733/81464890-1cd72480-91f8-11ea-83d3-e92445500e09.jpg)

the attached file is full report.

[report20200508-2353.zip](https://github.com/AlexeyAB/darknet/files/4602748/report20200508-2353.zip)

**Comment by AlexeyAB** [comment]
Can you show such screeshot for another video file?

**Comment by goodtogood** [comment]
> detector demo ./A.data ./A.cfg A.weights A.mp4 -benchmark

![model A](https://user-images.githubusercontent.com/10088733/81473605-27161480-9232-11ea-9957-ba0d33e1376f.jpg)

**Comment by AlexeyAB** [comment]
Thanks! Also can you show such screenshot if you use Commit on Apr 17, 2020 `Added param train_only_bn=1 to train only Batch-normalization for thi…` https://github.com/AlexeyAB/darknet/tree/d4b2ed9d22210d33e438d9557e976f8053f1cf9b

https://github.com/AlexeyAB/darknet/archive/d4b2ed9d22210d33e438d9557e976f8053f1cf9b.zip

**Comment by goodtogood** [comment]
profile report baseed on the commit you specified.

there's an obvious difference.

> detector demo ./A.data ./A.cfg A.weights A.mp4 -benchmark

![profile](https://user-images.githubusercontent.com/10088733/81475799-00aba580-9241-11ea-93f5-cb7cdc0c9f73.jpg)

> detector demo ./B.data ./B.cfg B.weights B.mp4 -benchmark

![profile B](https://user-images.githubusercontent.com/10088733/81476009-19688b00-9242-11ea-9179-1ce6be35c1db.jpg)
